frontend/packages/kubevirt-plugin/src/components/boot-order/boot-order.tsx

```diff
@@ -0,0 +1,126 @@
+import * as React from 'react';
+import * as _ from 'lodash';
+import { Text, TextVariants } from '@patternfly/react-core';
+import { DNDDataList, DNDDataListItem } from '../dnd-list';
+import { BootOrderEmpty } from './boot-order-empty';
+import { AddItemButton } from './add-item-button';
+import { AddItemFormSelect } from './add-item-form-select';
+import {
+  addItemMessage,
+  addItemDisabledMessage,
+  addItemSelectLabel,
+  bootOrderEmptyMessage,
+  bootOrderEmptyTitle,
+  deviceKey,
+  deviceLabel,
+  bootOrderAriaLabel,
+} from './constants';
+
+export const BootOrder = ({ devices, setDevices }: BootOrderProps) => {
+  const sources = _.sortBy(devices.filter((device) => device.value.bootOrder), 'value.bootOrder');
+  const options = devices.filter((device) => !device.value.bootOrder);
+  const [isEditMode, setEditMode] = React.useState<boolean>(false);
+
+  // Add a bootOrder to a device by key, item key is "<type>->name>".
+  const addByKey = (key: string): void => {
+    const newDevices = _.cloneDeep(devices);
+
+    const bootOrder = sources.length !== 0 ? sources[sources.length - 1].value.bootOrder + 1 : 1;
+    newDevices.find((device) => deviceKey(device) === key).value.bootOrder = bootOrder;
+
+    setDevices(newDevices);
+    setEditMode(false);
+  };
+
+  // Remove a bootOrder from a device by index.
+  const onDelete = (index: number) => {
+    const newDevices = _.cloneDeep(devices);
+
+    const key = deviceKey(sources[index]);
+    delete newDevices.find((device) => deviceKey(device) === key).value.bootOrder;
+
+    // Reset newDevices bootOrder using the list index.
+    _.filter(newDevices, (device) => device.value.bootOrder).forEach((source, i) => {
```

Comment:
we might not need the _.cloneDeep if we use `devices.map` and `_.merge` instead

Reply:
I think `_,marge` creates a new object like `_.cloneDeep` does ?

Reply:
`_.merge({}, originalDevice, { value: { bootOrder: 3}})` will just create a shallow copy of the outer wrapper object  and shallow copy of value, but it should keep the other references intact. Thus it should be better optimized than cloneDeep which creates everything from scratch.